specgen/specgen/hellocap/listener.go

```diff
@@ -0,0 +1,200 @@
+}
+
+// attachHello is used to put a captured ClientHello on a http.Request context. This is how the
+// ClientHello, captured at the TCP level, is made available to users of this library.
+func attachHello(req *http.Request, hello capturedHello) *http.Request {
+	ctx := context.WithValue(req.Context(), contextHelloKey, hello)
+	return req.Clone(ctx)
+}
+
+// onHello is a callback invoked when a ClientHello is captured. Only one of hello or err will be
+// non-nil. This callback is not invoked for incomplete ClientHellos. In other words, err is non-nil
+// only if what has been read off the connection could not possibly constitute a ClientHello,
+// regardless of further data from the connection.
+type onHello func(hello []byte, err error)
+
+// capturingConn is a TCP connection and implements net.Conn. If a TLS ClientHello is sent by the
+// peer, capturingConn will invoke the provided onHello callback. This is the basic building block
+// by which we build a hello-capturing server.
+type capturingConn struct {
+	// Wraps a TCP connection.
+	net.Conn
+
+	helloRead bool
+	helloBuf  *bytes.Buffer
+	helloLock sync.Mutex // protects fields in this block
+
+	onHello onHello
+	onClose func()
+}
+
+func newCapturingConn(wrapped net.Conn, onHello onHello, onClose func()) *capturingConn {
+	return &capturingConn{
+		wrapped,
+		false,
+		new(bytes.Buffer),
+		sync.Mutex{},
+		onHello,
+		onClose,
+	}
+}
+
+func (c *capturingConn) Read(b []byte) (n int, err error) {
+	n, err = c.Conn.Read(b)
+	c.checkHello(b[:n])
+	return
+}
+
+func (c *capturingConn) checkHello(newBytes []byte) {
+	c.helloLock.Lock()
+	if !c.helloRead {
+		c.helloBuf.Write(newBytes)
+		if c.helloBuf.Len() >= tlsRecordHeaderLen {
+			helloLen, err := parseClientHelloHeader(c.helloBuf.Bytes()[:tlsRecordHeaderLen])
+			if err != nil {
+				c.onHello(nil, fmt.Errorf("failed to parse header: %w", err))
+				c.helloRead = true
+			}
+			if c.helloBuf.Len() >= helloLen {
+				c.onHello(c.helloBuf.Bytes()[:helloLen+tlsRecordHeaderLen], nil)
+				c.helloRead = true
+			}
+		}
+	}
+	c.helloLock.Unlock()
+}
+
+func (c *capturingConn) Close() error {
+	c.onClose()
+	return c.Conn.Close()
+}
+
+type capturedHello struct {
+	hello []byte
+	err   error
+}
+
+// handlerListener is used to listen for raw TCP connections as part of an HTTPS server. The TCP
+// connections are assumed to come from TLS clients. handlerListener captures the TLS ClientHello
+// (using capturingConn) and attaches it to the *http.Request context (using attachHello).
+//
+// Implements net.Listener and http.Handler.
+type handlerListener struct {
+	tcpListener net.Listener
+	handler     http.Handler
+
+	hellosByRemote map[string]capturedHello
+	sync.Mutex
+}
+
+func newHandlerListener(addr string, handler http.Handler) (*handlerListener, error) {
+	l, err := net.Listen("tcp", addr)
+	if err != nil {
+		return nil, err
+	}
+	return &handlerListener{
+		l, handler, map[string]capturedHello{}, sync.Mutex{},
```

Comment:
An empty mutex is the default value, and I would use explicit initialization here.

Reply:
> An empty mutex is the default value

That's correct.

> I would use explicit struct initialization here

I prefer this form (is that called implicit struct initialization? That sounds like a term I just made up).  This way, the compiler checks maintainers who add or remove fields.

Reply:
Huh, does `govet` not give you warnings on these? Mine does, and I never questioned it. I see your argument.